src/Raven.Server/Documents/Queries/WhereFieldTokens.cs

```diff
@@ -60,30 +60,118 @@ private static string ExtractTokenValue(ValueToken valueToken, string queryText)
 
                 queryParameters.GetPropertyByIndex(index, ref propertyDetails);
 
-                switch (propertyDetails.Token)
+                return GetValueAndTypeFromJson(propertyDetails);
+            }
+
+            return (ExtractTokenValue(valueToken, queryText), Type);
+        }
+
+        private (List<string>, ValueTokenType) ExtractInParameters(ValueToken valueToken, string queryText, BlittableJsonReaderObject queryParameters)
+        {
+            Debug.Assert(Type == ValueTokenType.Parameter);
+
+            var property = new BlittableJsonReaderObject.PropertyDetails();
+
+            var parameterName = QueryExpression.Extract(queryText, valueToken);
+
+            var index = queryParameters.GetPropertyIndex(parameterName);
+
+            queryParameters.GetPropertyByIndex(index, ref property);
+
+            switch (property.Token)
+            {
+                case BlittableJsonToken.StartArray:
+                case BlittableJsonToken.StartArray | BlittableJsonToken.OffsetSizeByte:
+                case BlittableJsonToken.StartArray | BlittableJsonToken.OffsetSizeShort:
+                    if (Operator != OperatorType.In)
+                        throw new InvalidOperationException("Array parameter is supported only as a parameter of IN operator");
+
+                    return GetFlatValuesFromJsonArray((BlittableJsonReaderArray)property.Value);
+                default:
+                    var valueType = GetValueAndTypeFromJson(property);
+                    return (new List<string>{valueType.Value}, valueType.Type);
+            }
+        }
+
+        private static (string Value, ValueTokenType Type) GetValueAndTypeFromJson(BlittableJsonReaderObject.PropertyDetails property)
```

Comment:
Just looked deeper and see that we are parsing strings back to numeric values. For example, `long` handling in range query:

https://github.com/ravendb/ravendb/blob/v4.0/src/Raven.Server/Documents/Queries/Parse/LuceneAST.cs#L448

https://github.com/ravendb/ravendb/blob/v4.0/src/Raven.Server/Documents/Queries/Parse/LuceneAST.cs#L461

We'll change that to avoid `number -> string -> number ` conversions